### PR TITLE
Add new mode to get margin for SR anomaly detector

### PR DESCRIPTION
**Pull request by mengaims**
#4958 We are excited to review your PR.

So we can do the best job, please check:

- [ ] There's a descriptive title that will make sense to other developers some time from now. 
- [ ] There's associated issues. All PR's should have issue(s) associated - unless a trivial self-evident change such as fixing a typo. You can use the format `Fixes #nnnn` in your description to cause GitHub to automatically close the issue(s) when your PR is merged.
- [ ] Your change description explains what the change does, why you chose your approach, and anything else that reviewers should know.
- [ ] You have included any necessary tests in the same PR.

### CODE REVIEW

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/TimeSeries/DetectAnomalyAndMarginBySrCnn.cs [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+            //5       0       0.00            0.18    5.00            5.00            5.01            4.99
+            //5       0       0.00            0.18    5.00            5.00            5.01            4.99
+            //5       0       0.00            0.18    5.00            5.00            5.01            4.99
+            //5       0       0.00            0.18    5.00            5.00            5.01            4.99
+            //5       0       0.00            0.18    5.00            5.00            5.01            4.99
+
+            // Anomaly.
+            PrintPrediction(10, engine.Predict(new TimeSeriesData(10)));
+
+            //10      1       0.38            0.93    9.06            5.00            9.07            9.05    <-- alert is on, predicted anomaly
```
[the remark this answers is not in the document]
Comment (mengaims) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/TimeSeries/DetectAnomalyAndMarginBySrCnn.cs]: Fixed

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/TimeSeries/DetectAnomalyAndMarginBySrCnnBatchPrediction.cs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            //5       0       0.00            0.00    0.00            0.00            0.00            0.00
+            //5       0       0.00            0.00    0.00            0.00            0.00            0.00
+            //5       0       0.00            0.00    0.00            0.00            0.00            0.00
+            //5       0       0.00            0.00    0.00            0.00            0.00            0.00
+            //5       0       0.00            0.18    5.00            5.00            5.01            4.99
+            //5       0       0.00            0.18    5.00            5.00            5.01            4.99
+            //5       0       0.00            0.18    5.00            5.00            5.01            4.99
+            //5       0       0.00            0.18    5.00            5.00            5.01            4.99
+            //5       0       0.00            0.18    5.00            5.00            5.01            4.99
+            //10      1       0.38            0.93    9.06            5.00            9.07            9.05
```
[the remark this answers is not in the document]
Comment (mengaims) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/TimeSeries/DetectAnomalyAndMarginBySrCnnBatchPrediction.cs]: Fixed

File: src/Microsoft.ML.TimeSeries/SrCnnAnomalyDetectionBase.cs [hunk 3]
```diff
@@ -292,6 +358,282 @@ private Single CalculateSocre(Single mag, Single avgMag)
                     }
                     return (float)(Math.Abs(mag - avgMag) / safeDivisor);
                 }
+
+                private List<int> GetAnomalyIndex(List<Single> magList, List<Single> avgMagList)
+                {
+                    List<int> anomalyIdxList = new List<int>();
+                    for (var i = 0; i < magList.Count; ++i)
+                    {
+                        double safeDivisor = avgMagList[i];
```
Comment (yaeldekel) [inline src/Microsoft.ML.TimeSeries/SrCnnAnomalyDetectionBase.cs]: >avgMagList[i] [](start = 45, length = 13)

Can this be negative? If yes, then you need to check the absolute value of `safeDivisor`.

File: src/Microsoft.ML.TimeSeries/SrCnnAnomalyDetectionBase.cs [hunk 4]
```diff
@@ -292,6 +358,282 @@ private Single CalculateSocre(Single mag, Single avgMag)
...
+                        if (safeDivisor < 1e-8)
+                        {
+                            safeDivisor = 1e-8;
+                        }
+                        var score = (Math.Abs((double)magList[i] - (double)avgMagList[i]) / safeDivisor);
+                        if (score > 3.5f)
```
Comment (yaeldekel) [inline src/Microsoft.ML.TimeSeries/SrCnnAnomalyDetectionBase.cs]: >3.5f [](start = 36, length = 4)

What is the significance of this value? Can you add a comment, or define this as a constant somewhere?

File: src/Microsoft.ML.TimeSeries/SrCnnAnomalyDetectionBase.cs [hunk 5]
```diff
@@ -292,6 +358,282 @@ private Single CalculateSocre(Single mag, Single avgMag)
...
+                            anomalyIdxList.Add(i);
+                        }
+                    }
+
+                    return anomalyIdxList;
+                }
+
+                private List<Double> GetDeanomalyData(List<Double> data, List<int> anomalyIdxList)
+                {
+                    //Console.WriteLine("data: " + string.Join(",", data));
```
Comment (yaeldekel) [inline src/Microsoft.ML.TimeSeries/SrCnnAnomalyDetectionBase.cs]: >//Console.WriteLine("data: " + string.Join(",", data)); [](start = 20, length = 55)

Delete.

File: src/Microsoft.ML.TimeSeries/SrCnnAnomalyDetectionBase.cs [hunk 6]
```diff
@@ -6,6 +6,8 @@
 using System.Collections.Generic;
 using System.IO;
 using System.Linq;
+using System.Text;
```
Comment (yaeldekel) [inline src/Microsoft.ML.TimeSeries/SrCnnAnomalyDetectionBase.cs]: Is this needed?

File: src/Microsoft.ML.TimeSeries/SRCNNAnomalyDetector.cs [hunk 7]
```diff
@@ -95,15 +106,17 @@ public SrCnnArgument(SrCnnAnomalyDetector transform)
...
             }
         }
 
         private static VersionInfo GetVersionInfo()
         {
             return new VersionInfo(
                 modelSignature: "SRCNTRNS",
-                verWrittenCur: 0x00010001, // Initial
-                verReadableCur: 0x00010001,
+                verWrittenCur: 0x00010010, // Initial
```
Comment (yaeldekel) [inline src/Microsoft.ML.TimeSeries/SRCNNAnomalyDetector.cs]: >0x00010010 [](start = 31, length = 10)

0x00010002

Comment (yaeldekel) [inline src/Microsoft.ML.TimeSeries/SRCNNAnomalyDetector.cs]: >Initial [](start = 46, length = 7)

Please change this to explain what caused the version change.

Comment (yaeldekel) [reply src/Microsoft.ML.TimeSeries/SRCNNAnomalyDetector.cs]: Actually, where is the change in the way the model is serialized? I don't see any change to the serialization/deserialization methods in this file.

---
In reply to: [397223419](https://github.com/dotnet/machinelearning/pull/4946#discussion_r397223419) [](ancestors = 397223419)

File: src/Microsoft.ML.TimeSeries/SrCnnTransformBase.cs [hunk 8]
```diff
@@ -49,6 +56,14 @@ internal abstract class SrCnnArgumentBase
         [Argument(ArgumentType.AtMostOnce, HelpText = "The threshold to determine anomaly, score larger than the threshold is considered as anomaly.",
             ShortName = "thre", SortOrder = 9)]
         public double Threshold = 0.3;
+
+        [Argument(ArgumentType.AtMostOnce, HelpText = "The detection mode, affect output vector length",
```
Comment (ganik) [inline src/Microsoft.ML.TimeSeries/SrCnnTransformBase.cs]: >affect [](start = 75, length = 6)

affects

Comment (ganik) [inline src/Microsoft.ML.TimeSeries/SrCnnTransformBase.cs]: >h" [](start = 101, length = 2)

Add .

File: src/Microsoft.ML.TimeSeries/SrCnnTransformBase.cs [hunk 9]
```diff
@@ -49,6 +56,14 @@ internal abstract class SrCnnArgumentBase
...
+            ShortName = "mode", SortOrder = 10)]
+        public SrCnnDetectMode SrCnnDetectMode = SrCnnDetectMode.AnomalyOnly;
+
+        [Argument(ArgumentType.AtMostOnce, HelpText = "The sensitivity of boundary",
```
Comment (ganik) [inline src/Microsoft.ML.TimeSeries/SrCnnTransformBase.cs]: >y" [](start = 81, length = 2)

Same, add .

**Comment by harishsk** [comment]
@mengaims Can this PR be closed?

**Comment by harishsk** [comment]
Closing because this is superseded by #4998

**Pull request closed without merging**